File: 1-two-sum/1-two-sum.py
class Solution(object):
    def twoSum(self, nums, target):
        #Brute Force Method:
        # We could go through the entire array in O(n) time and for each 
        # number, we could do requiredNum = (target-thatNumber) and traverse
        # through the rest of the array and try to find the requireNum
        # If that number exists, we return the index of both
        #Efficient Approach:
        # We could go through the entire array in O(n) time and if that value
        # doesn't exist in the keys of a dictionary, then we add it as a key
        # and the index as the value
        # If the key exists, then we append the index value to the value array
        # Once this has been done, we could go through the dictionary key
        # by key and subtract that key from the target and if the
        # result = target - currentKey exists in the keys, then we find it
        # and output the index
        hashMap = {}
        for i,num in enumerate(nums):
            if(num not in hashMap.keys()):
                hashMap[num] = [i]
            else:
                hashMap.get(num).append(i)
        indices = []
        for key,value in hashMap.items():
            differenceRequired = target - key
            index = value.pop()
            if((differenceRequired in hashMap.keys()) and (hashMap.get(differenceRequired)!=[])):
                indices.append(index)
                indices.append(hashMap.get(differenceRequired).pop())
                break
            # value.append(index)
        return indices

File: 1-two-sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {
    /* 
    Brute Force Method:
    We could just iterate through the entire array element-by-element and for each element, we check
    whether the target - that element exists and we return the first pair of those elements
    This would give is a time complexity of O(n^2)

    Efficient Method:
    Example: 
        nums = [2,7,11,15]
        target = 9
    Expected = [0,1] so 2+7
    What we can do is that we can use a dictionary
    We can basically iterate through the entire array one by one and in the dictionary, we can have the key be the 
    difference (target - current element) and the value be the index of the curent element
    So for example, the dictionary will be = {7:0, 2:1, -2:2, -6:3}
    As we go through the array, we keep checking whether they key that is the current element exists and if it does, then
    we output an array of the value of that key, and the current index.
    If it doesn't exist, then we move onto the next element in the array
        */

    const dict = {};
    const len = nums.length;
    for(let i=0; i<len; i++){
        if(nums[i] in dict){
            return Array.of(dict[nums[i]],i);
        }
        dict[target-nums[i]] = i
    }
};
